### PR TITLE
move uploader version check before schema validation

### PR DESCRIPTION
**Pull request by jebeck**
What it says on the tin. This is what we talked about @jh-bate, complete with a comment and a ¯\_(ツ)_/¯ emoji

I even added a new type of API ingestion test to make sure it will send back the proper error on an attempt to upload metadata with outdated version.

### CODE REVIEW

**Comment by jh-bate** [comment]
LGTM @darinkrauss any comments? both @jebeck and I thought it might be the easiest way forward .... but it is another compromise

File: lib/dataBroker.js [hunk 1]
```diff
@@ -39,6 +40,28 @@ module.exports = function(streamDAO) {
...
+       * This is a hacky place to move this logic, but we need it to run
+       * before schema validation is attempted (or else it's pointless)
+       * so ¯\_(ツ)_/¯
+       */
+      if (datum.type === 'upload') {
+        var versionStr = datum.version.toLowerCase();
+
+        // TODO: longterm this check should be against a datamodel version
+        // and also probably not a hard go/no-go check
+        if (versionStr.indexOf('tidepool-uploader') !== -1 && uploadConfig.minimumUploaderVersion !== null ) {
```
Comment (darinkrauss) [inline lib/dataBroker.js]: I would put the check for `uploadConfig.minimumUploaderVersion != null` around the whole block.  If we don't have a minimumUploaderVersion from the config then don't do any of it.

Comment (darinkrauss) [inline lib/dataBroker.js]: When would `tidepool-uploader` not be in the version string?

**Comment by darinkrauss** [comment]
A couple of comments, but otherwise looks like.  I'd imagine we'll want to eventually do a check outside of this logic.  Perhaps have the uploader send something that is immediately checked upon the API invocation (well before the data broker).  :shipit:

File: lib/dataBroker.js (hunk 1 above)
Comment (jebeck) [inline lib/dataBroker.js]: Yeah, good idea on the check for `uploadConfig.minimumUploaderVersion` - doing that and throwing a different error if there's no configured version (as that's also a situation we don't want to be in).

`tidepool-uploader` would not be in the version string if it's 3rd-party devs uploading data, or at least that was the original thought.

**Comment by jebeck** [comment]
Yeah @darinkrauss we definitely want this version checking to precede even reading data from devices eventually, but we didn't want to hold up bootstrapping for it.

**Pull request merged**